pydm/widgets/rules.py

```diff
@@ -224,6 +227,38 @@ def run(self):
                         self.calculate_expression(widget_ref, idx, rule)
             self.msleep(33)  # 30Hz
 
+    def callback_enum(self, widget_ref, index, ch_index, enums):
+        """
+        Callback executed when a channel receives a new enum_string.
+
+        Parameters
+        ----------
+        widget_ref : weakref
+            A weakref to the widget owner of the rule.
+        index : int
+            The index of the rule being processed.
+        ch_index : int
+            The channel index on the list for this rule.
+        trigger : bool
+            Whether or not this channel should trigger a calculation of the
+            expression
+        value : any
+            The new value for this channel.
+
+        Returns
+        -------
+        None
+        """
+        try:
+            w_map = self.widget_map[widget_ref]
+            w_map[index]['enums'][ch_index] = enums
```

Comment:
If there's any way this can make a new key and mutate the dictionary, I think you might need a lock around this block. Haven't looked into the details to see if that's possible, though.

Reply:
The dictionary entries are pre-allocated so no risk of mutation on this point.
It was done this way to avoid locking on the callbacks. The create and delete of rules is properly locked.